src/NUnitEngine/nunit.engine/Services/ExtensionService.cs

```diff
@@ -0,0 +1,154 @@
+{
+    public class ExtensionService : Service
+    {
+        private List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
+        private Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
+        private Dictionary<Type, ExtensionPoint> _typeIndex = new Dictionary<Type, ExtensionPoint>();
+
+        private List<ExtensionNode> _extensions = new List<ExtensionNode>();
+
+        #region Public Properties
+
+        public IEnumerable<ExtensionPoint> ExtensionPoints
+        {
+            get { return _extensionPoints; }
+        }
+
+        public IEnumerable<ExtensionNode> Extensions
+        {
+            get { return _extensions;  }
+        }
+
+        #endregion
+
+        public ExtensionPoint GetExtensionPoint(string path)
+        {
+            return _pathIndex[path];
+        }
+
+        public ExtensionPoint GetExtensionPoint(Type type)
+        {
+            return _typeIndex[type];
+        }
+
+        public T[] GetExtensions<T>()
+        {
+            var extensions = new List<T>();
+
+            var ep = GetExtensionPoint(typeof(T));
+            foreach (var node in ep.Extensions)
+                extensions.Add((T)node.ExtensionObject);
+
+            return extensions.ToArray();
+        }
+
+        #region Service Overrides
+
+        public override void StartService()
+        {
+            var thisAssembly = Assembly.GetExecutingAssembly();
+            var startDir = new DirectoryInfo(AssemblyHelper.GetDirectoryName(thisAssembly));
+
+            FindExtensionPoints(thisAssembly);
+            FindExtensionsInDirectory(startDir);
+
+            base.StartService();
+        }
+
+        private void FindExtensionPoints(Assembly assembly)
+        {
+            foreach (ExtensionPointAttribute attr in assembly.GetCustomAttributes(typeof(ExtensionPointAttribute), false))
+            {
+                var ep = new ExtensionPoint(attr.Path, attr.Type)
+                {
+                    Description = attr.Description
+                };
+
+                _extensionPoints.Add(ep);
+
+                // TODO: These Adds will throw if the path or type is a duplicate.
+                // We should give an error message instead.
+                _pathIndex.Add(ep.Path, ep);
+                _typeIndex.Add(ep.Type, ep);
+            }
+        }
+
+        private void FindExtensionsInDirectory(DirectoryInfo startDir)
+        {
+            var addinsFiles = startDir.GetFiles("*.addins");
+            if (addinsFiles.Length > 0)
+                foreach (var file in addinsFiles)
+                    ProcessAddinsFile(startDir.FullName, file.FullName);
+            else
+                foreach (var file in startDir.GetFiles("*.dll"))
+                    FindExtensionsInAssembly(file.FullName);
+        }
+
+        private void ProcessAddinsFile(string baseDir, string fileName)
+        {
+            var doc = new XmlDocument();
```

Comment:
I planned on a flat file (see spec) but then adjusted so I could use the
pre-existing file from mono.addins. Their format has directory entries and
patterns/names of assemblies to ignore when scanning. Up to now, I have not
felt the need for the ignored files.

My own plan was to have lines for directories and lines for files, which
might include wild-cards. It seems like it may be easier to just indicate
which lines are directories and which are files rather than try to figure
it out. We could use a prefix like D: or A: for directories and assemblies,
but XML seemed pretty easy for this.

Bottom line: I think we can go either way. What do you think, considering
the possible addition of individual assemblies and bearing in mind that
users may edit this file.

On Fri, Aug 21, 2015 at 10:58 AM, Rob Prouse <notifications@github.com>
wrote:

> In src/NUnitEngine/nunit.engine/Services/ExtensionService.cs
> <https://github.com/nunit/nunit/pull/774#discussion_r37660194>:
>
> > +        }
> > +
> > +        private void FindExtensionsInDirectory(DirectoryInfo startDir)
> > +        {
> > +            var addinsFiles = startDir.GetFiles("*.addins");
> > +            if (addinsFiles.Length > 0)
> > +                foreach (var file in addinsFiles)
> > +                    ProcessAddinsFile(startDir.FullName, file.FullName);
> > +            else
> > +                foreach (var file in startDir.GetFiles("*.dll"))
> > +                    FindExtensionsInAssembly(file.FullName);
> > +        }
> > +
> > +        private void ProcessAddinsFile(string baseDir, string fileName)
> > +        {
> > +            var doc = new XmlDocument();
>
> Using XML for this file seems like overkill for a list of addin
> directories. Wouldn't it be easier to just have a flat file, one addin
> directory per line? We could ignore blank lines and lines starting with #.
> Were you planning on including other information in the addins files?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/pull/774/files#r37660194>.
>